src/ofonotracker.cpp

```diff
@@ -0,0 +1,335 @@
+ * ========================================================================= */
+
+struct ModemInfo
+{
+    ModemInfo(QString path, const QVariantMap &properties)
+        : m_objectPath(path)
+        , m_properties(properties)
+    {
+    }
+    ModemInfo(QString path)
+        : m_objectPath(path)
+    {
+    }
+    ModemInfo(const ModemInfo &that)
+        : m_objectPath(that.m_objectPath)
+        , m_properties(that.m_properties)
+    {
+    }
+    ModemInfo()
+    {
+    }
+    QDBusObjectPath m_objectPath;
+    QVariantMap m_properties;
+};
+Q_DECLARE_METATYPE(ModemInfo)
+
+QDBusArgument &operator<<(QDBusArgument &argument, const ModemInfo &modemInfo)
+{
+    argument.beginStructure();
+    argument << modemInfo.m_objectPath;
+    argument << modemInfo.m_properties;
+    argument.endStructure();
+    return argument;
+}
+const QDBusArgument &operator>>(const QDBusArgument &argument, ModemInfo &modemInfo)
+{
+    argument.beginStructure();
+    argument >> modemInfo.m_objectPath;
+    argument >> modemInfo.m_properties;
+    argument.endStructure();
+    return argument;
+}
+
+typedef QList<ModemInfo> ModemInfoList;
+Q_DECLARE_METATYPE(ModemInfoList)
+
+static void registerDBusTypes()
+{
+    static bool registered(false);
+    if (!registered) {
+        registered = true;
+        qDBusRegisterMetaType<ModemInfo>();
+        qDBusRegisterMetaType<ModemInfoList>();
+    }
+}
+
+/* ========================================================================= *
+ * OFonoTrackerPrivate
+ * ========================================================================= */
+
+class OFonoTrackerPrivate : public QObject, protected QDBusContext
+{
+    Q_OBJECT;
+public:
+    OFonoTrackerPrivate(bool synchronousInit, OFonoTracker *parent);
+    virtual ~OFonoTrackerPrivate();
+    QStringList imeiNumbers() const;
+private Q_SLOTS:
+    void evaluateProperties();
+    void handleManagerModemAdded(QDBusObjectPath objectPath, QVariantMap properties);
+    void handleManagerModemRemoved(QDBusObjectPath objectPath);
+    void handleModemPropertyChanged(QString key, QDBusVariant value);
+private:
+    void evaluatePropertiesLater();
+    void getModemsAsync();
+    void getModemsSync();
+    void flushModems();
+    void setModems(const ModemInfoList &modemInfoList);
+    void addModem(const ModemInfo &modemInfo);
+    void removeModem(QString path);
+    void setModemProperty(QString path, QString key, QVariant val);
+    static bool trackedModemProperty(const QString &key);
+    OFonoTracker *q_ptr;
+    QDBusConnection m_systemBus;
+    QStringList m_modemPathList;
+    QHash<QString, ModemInfo> m_modemInfoHash;
+    QTimer *m_evaluatePropertiesTimer;
+    QStringList m_imeiNumbers;
+    Q_DISABLE_COPY(OFonoTrackerPrivate);
+    Q_DECLARE_PUBLIC(OFonoTracker);
+};
+
+OFonoTrackerPrivate::OFonoTrackerPrivate(bool synchronousInit, OFonoTracker *parent)
+    : QObject(parent)
+    , q_ptr(parent)
+    , m_systemBus(QDBusConnection::systemBus())
```

Comment:
Or could just use the ::systemBus() where needed.

Reply:
That feels so wrong, but I guess those connected signals == even without outside influence that connection does not get dropped? Or is systemBus() something that is never closed?

Reply:
Due to use of libqofono: not relevant anymore